File: src/main/java/kr/dogfoot/hwpxlib/object/content/context_hpf/ContentHPFFile.java
package kr.dogfoot.hwpxlib.object.content.context_hpf;

import kr.dogfoot.hwpxlib.object.common.ObjectList;
import kr.dogfoot.hwpxlib.object.common.ObjectType;
import kr.dogfoot.hwpxlib.object.common.SwitchableObject;

public class ContentHPFFile extends SwitchableObject {
    private String version;
    private String uniqueIdentifier;
    private String id;
    private MetaData metaData;
    private ObjectList<ManifestItem> manifest;
    private ObjectList<SpineItemRef> spine;

    public ContentHPFFile() {
    }

    @Override
    public ObjectType _objectType() {
        return ObjectType.opf_package;
    }


    public String version() {
        return version;
    }

    public void version(String version) {
        this.version = version;
    }

    public ContentHPFFile versionAnd(String version) {
        this.version = version;
        return this;
    }

    public String uniqueIdentifier() {
        return uniqueIdentifier;
    }

    public void uniqueIdentifier(String uniqueIdentifier) {
        this.uniqueIdentifier = uniqueIdentifier;
    }

    public ContentHPFFile uniqueIdentifierAnd(String uniqueIdentifier) {
        this.uniqueIdentifier = uniqueIdentifier;
        return this;
    }

    public String id() {
        return id;
    }

    public void id(String id) {
        this.id = id;
    }

    public ContentHPFFile idAnd(String id) {
        this.id = id;
        return this;
    }

    public MetaData metaData() {
        return metaData;
    }

    public void createMetaData() {
        metaData = new MetaData();
    }

    public void removeMetaData() {
        metaData = null;
    }

    public ObjectList<ManifestItem> manifest() {
        return manifest;
    }

    public void createManifest() {
        manifest = new ObjectList<ManifestItem>(ObjectType.opf_manifest, ManifestItem.class);
    }

    public void removeManifest() {
        manifest = null;
    }

    public ManifestItem getManifestItemById(String id) {
        for (ManifestItem item : manifest.items()) {
            if (item.id().equals(id)) {
                return item;
            }
        }
        return null;
    }

    public ObjectList<SpineItemRef> spine() {
        return spine;
    }

    public void createSpine() {
        spine = new ObjectList<SpineItemRef>(ObjectType.opf_spine, SpineItemRef.class);
    }

    public void removeSpine() {
        spine = null;
    }

    @Override
    public ContentHPFFile clone() {
        ContentHPFFile cloned = new ContentHPFFile();
        cloned.copyFrom(this);
        return cloned;
    }

    public void copyFrom(ContentHPFFile from) {
        this.version = from.version;
        this.uniqueIdentifier = from.uniqueIdentifier;
        this.id = from.id;

        if (from.metaData != null) {
            this.metaData = from.metaData.clone();
        } else {
            this.metaData = null;
        }

        if (from.manifest != null) {
            createManifest();
            for (ManifestItem manifestItem : from.manifest.items()) {
                manifest.add(manifestItem.clone());
            }
        } else {
            removeManifest();
        }

        if (from.spine != null) {
            createSpine();
            for (SpineItemRef spineItemRef : from.spine.items()) {
                spine.add(spineItemRef.clone());
            }
        } else {
            removeSpine();
        }

        super.copyFrom(from);
    }
}

File: js/src/object/content/context_hpf/ContentHPFFile.js
import { SwitchableObject } from '../../common/SwitchableObject.js';
import { ObjectType } from '../../common/ObjectType.js';

/**
 * Manifest item in content.hpf
 */
export class ManifestItem {
  constructor() {
    this._id = null;
    this._href = null;
    this._mediaType = null;
    this._attachedFile = null;
  }

  _objectType() {
    return ObjectType.opf_item;
  }

  id() { return this._id; }
  idSet(value) { this._id = value; }
  idAnd(value) { this._id = value; return this; }

  href() { return this._href; }
  hrefSet(value) { this._href = value; }
  hrefAnd(value) { this._href = value; return this; }

  mediaType() { return this._mediaType; }
  mediaTypeSet(value) { this._mediaType = value; }
  mediaTypeAnd(value) { this._mediaType = value; return this; }

  hasAttachedFile() {
    return this._mediaType && (
      this._mediaType.startsWith('image/') ||
      this._mediaType === 'application/ole' ||
      this._mediaType.startsWith('application/x-javascript')
    );
  }

  attachedFile() { return this._attachedFile; }

  createAttachedFile() {
    this._attachedFile = { data: null };
    return this._attachedFile;
  }

  clone() {
    const cloned = new ManifestItem();
    cloned._id = this._id;
    cloned._href = this._href;
    cloned._mediaType = this._mediaType;
    if (this._attachedFile) {
      cloned._attachedFile = { data: this._attachedFile.data };
    }
    return cloned;
  }
}

/**
 * Manifest in content.hpf
 */
export class Manifest {
  constructor() {
    this._items = [];
  }

  _objectType() {
    return ObjectType.opf_manifest;
  }

  items() {
    return this._items;
  }

  addItem() {
    const item = new ManifestItem();
    this._items.push(item);
    return item;
  }

  clone() {
    const cloned = new Manifest();
    for (const item of this._items) {
      cloned._items.push(item.clone());
    }
    return cloned;
  }
}

/**
 * Spine item ref
 */
export class SpineItemRef {
  constructor() {
    this._idref = null;
  }

  idref() { return this._idref; }
  idrefSet(value) { this._idref = value; }
  idrefAnd(value) { this._idref = value; return this; }

  clone() {
    const cloned = new SpineItemRef();
    cloned._idref = this._idref;
    return cloned;
  }
}

/**
 * Spine in content.hpf
 */
export class Spine {
  constructor() {
    this._itemRefs = [];
  }

  itemRefs() {
    return this._itemRefs;
  }

  addItemRef() {
    const ref = new SpineItemRef();
    this._itemRefs.push(ref);
    return ref;
  }

  clone() {
    const cloned = new Spine();
    for (const ref of this._itemRefs) {
      cloned._itemRefs.push(ref.clone());
    }
    return cloned;
  }
}

/**
 * Contents/content.hpf file representation
 */
export class ContentHPFFile extends SwitchableObject {
  constructor() {
    super();
    this._version = null;
    this._uniqueIdentifier = null;
    this._title = null;
    this._language = null;
    this._manifest = null;
    this._spine = null;
  }

  _objectType() {
    return ObjectType.opf_package;
  }

  version() { return this._version; }
  versionSet(value) { this._version = value; }

  uniqueIdentifier() { return this._uniqueIdentifier; }
  uniqueIdentifierSet(value) { this._uniqueIdentifier = value; }

  title() { return this._title; }
  titleSet(value) { this._title = value; }

  language() { return this._language; }
  languageSet(value) { this._language = value; }

  manifest() { return this._manifest; }

  createManifest() {
    this._manifest = new Manifest();
    return this._manifest;
  }

  spine() { return this._spine; }

  createSpine() {
    this._spine = new Spine();
    return this._spine;
  }

  clone() {
    const cloned = new ContentHPFFile();
    cloned.copyFrom(this);
    return cloned;
  }

  copyFrom(from) {
    this._version = from._version;
    this._uniqueIdentifier = from._uniqueIdentifier;
    this._title = from._title;
    this._language = from._language;
    if (from._manifest) {
      this._manifest = from._manifest.clone();
    }
    if (from._spine) {
      this._spine = from._spine.clone();
    }
    this._copyFromSwitchable(from);
  }
}

export default ContentHPFFile;
